glean-core/src/upload/util.rs

```diff
@@ -0,0 +1,262 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Pings directory processing utilities.
+
+use std::cmp::Ordering;
+use std::fs::{metadata, remove_file, File};
+use std::io::{BufRead, BufReader};
+use std::path::{Path, PathBuf};
+
+use lazy_static::lazy_static;
+use log;
+use regex::Regex;
+use serde_json::Value as JsonValue;
+
+use super::PingRequest;
+use crate::Result;
+
+lazy_static! {
+    /// Ping file names are UUIDs, this matches UUIDs for filtering purposes.
+    static ref PING_FILE_NAME_PATTERN: Regex = Regex::new("[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}").unwrap();
+}
+
+/// Attempt to delete a file.
+///
+/// This won't panic if not able to delete, just log.
+fn delete_file(path: &Path) {
+    match remove_file(path) {
+        Err(e) => log::error!("Error deleting file {}. {}", path.display(), e),
+        _ => log::info!("Deleted file {}", path.display()),
+    };
+}
+
+/// Get the file name from a path as a &str.
+///
+/// This won't panic if not able to get file name, just log.
+fn get_file_name_as_str(path: &Path) -> Option<&str> {
+    match path.file_name() {
+        None => {
+            log::warn!("Error getting file name from path: {}", path.display());
+            None
+        }
+        Some(file_name) => {
+            let file_name = file_name.to_str();
+            if file_name.is_none() {
+                log::warn!("File name is not valid unicode: {}", path.display());
+            }
+            file_name
+        }
+    }
+}
+
+/// Get the pending pings directory path.
+fn get_pings_dir(data_path: &Path) -> PathBuf {
+    data_path.join("pending_pings")
+}
+
+/// Reads a ping file and returns a `PingRequest` from it.
+///
+/// If the file is not properly formatted, it will be deleted.
+///
+/// ## Panics
+///
+/// Will panic if unable to read the file.
+fn process_ping_file(uuid: &str, path: &Path) -> Option<PingRequest> {
+    let file = File::open(path).expect("Should be able to read ping file.");
+    let mut lines = BufReader::new(file).lines();
+    // The way the ping file is structured, first line
+    // should always have the url and second line should
+    // have the body with the ping contents in JSON format
+    if let (Some(Ok(url)), Some(Ok(body))) = (lines.next(), lines.next()) {
+        if let Ok(parsed_body) = serde_json::from_str::<JsonValue>(&body) {
+            return Some(PingRequest::new(uuid, &url, parsed_body));
+        } else {
+            log::warn!("Can't parse ping contents as JSON.");
+        }
+    } else {
+        log::warn!("Ping file is not formatted as expected.");
+    }
+    delete_file(path);
```

Comment:
I don't think we actually want to delete the file until we know for sure it has been sent.
This is the only persistence we have for those.

Reply:
I delete only if the file is not formatted properly or if ping contents are not valid JSON. I also delete on the `process_pings_dir` function in case the file name is not a valid uuid.

I will send a different PR after this one to add the functionality where the upload module also receives the response from the upload and decides if should delete the file or try again.

Reply:
Oooh, sorry. Indeed. All good then.